81 avro (#165)
Accenture/reactive-interaction-gateway: docs/event-serializing.md

---

> id: event-serializing

Reviewer: ```suggestion
id: event-serialization
```

> title: Event Serializing

Reviewer: ```suggestion
title: Event Serialization
```

> sidebar_label: Event Serializing

Reviewer: ```suggestion
sidebar_label: Event Serialization
```

---

> We support [Apache Avro](https://avro.apache.org/) and standard JSON (de)serialization of events for producer as well as for consumer. This is working together with [Cloud Events specification](https://github.com/cloudevents/spec) (versions `0.1` and `0.2`). To make both of these worlds work we are utilizing Kafka headers introduced in version `0.11`. As Cloud Events specify there are 2 modes how to make this combination work. Discussion about this topic can be followed on [Github](https://github.com/cloudevents/spec/pull/337/files).

Reviewer: ```suggestion
We support [Apache Avro](https://avro.apache.org/) and standard JSON (de)serialization of events for both Kafka consumers and producers. We aim at following the [CloudEvents specification](https://github.com/cloudevents/spec) as closely as possible (the related part of the spec is called _transport bindings_ and is, at the time of writing, still under development). For the Kafka transport binding, we utilize Kafka headers introduced in Kafka version 0.11. CloudEvents currently defines two modes of operation; the related discussion can be followed on [Github](https://github.com/cloudevents/spec/pull/337/files).
```

**Structured content mode**

> This mode should be used when JSON format is desired, thus no further serialization and deserialization such as Apache Avro.

Reviewer: not sure I get this 😕 

1. if you need json, you can use both modes, the difference is just that in structured mode the context attributes are included in the json too.
1. you can use structured mode with Avro too, right?
1. IIRC we wanted to default to structured mode and JSON encoding - is this also what's implemented? If it is, you should probably mention that here..

Author: 1. I'll reword it
2,3. check `Producer details` and `Consumer details` they should provide enough information how it works -- if not, need change

> ##### KAFKA EVENT HEADERS
>
> ```

> > cloudEvent_contentType: "application/json"

Reviewer: shouldn't we use the newer spec here, e.g. `ce-contenttype`?

Author: is there already a newer? crap ...

Author: Where did you find it?

> ```
>
> ##### KAFKA EVENT VALUE
>
> ```
> {
>    "specversion" : "0.1",
>    "type" : "com.example.someevent",
>    "data": {},
>    ...
> }
> ```

**Binary content mode**

> This mode should be used when further serialization and deserialization is desired (e.g. Apache Avro).

Reviewer: again a bit unclear, probably due to what I hint on with my comment above

Author: Will be reworded.

> ##### KAFKA EVENT HEADERS
>
> ```
> cloudEvent_contentType: "avro/binary"
> cloudEvent_specversion: "0.2"

> > cloudEvent_type: "com.example.someevent"

Reviewer: again, we should only use the newer versions here (e.g., ce-contenttype, ce-specversion, ce-type). If we support both that's fine, but in the examples I feel we should be current.

Author: Can you provide source where they changed it? Checked PR that's referenced in docs and github, but no clear visibility, or I'm blind.

> ...
> ```
>
> ##### KAFKA EVENT VALUE
>
> ```
> ... application data ...
> ```

## Apache Avro format

> To be compatible across the board Apache Avro is using specific [format](https://docs.confluent.io/current/schema-registry/docs/serializer-formatter.html#wire-format) that needs to be followed.

Reviewer: isn't that like the main purpose of a format? :D

Author: I don't follow.

Reviewer: well Avro is a format specification, so in order to be compatible with Avro - not surprisingly - you gotta follow the format spec. Just thought it's funny, no worries ;)

```bash
# 0 - magic byte
# 1-4 - schema id - this is used by consumer to know which schema to use for deserialization
# 5-... - data

# Example
<0, 0, 0, 0, 1, 5, 3, 8, ...>
```

### Setup

![event-serializing-avro](assets/event-serializing-avro.png)

> Setup for event (de)serialization if fairly straightforward. Besides classic components like web client, RIG and Kafka it needs also `Kafka Schema Registry`. `Kafka Schema Registry` is place where all schemas are stored and retrieved from. As event is consumed from Kafka, RIG fetches schema, caches it in-memory and does deserialization. After that event is sent to web client (or any other client able to connect via WS/SSE). As for producing, RIG fetches schema, caches it in-memory and serialize event. Then it's sent to Kafka.

Reviewer: ```suggestion
Adopting Avro for event (de)serialization is fairly straightforward. First you need to run an instance of the `Kafka Schema Registry`, which is a central store for all Avro schemas in use. As an event is consumed from Kafka, RIG fetches the corresponding schema from the registry and deserializes the event with it, caching the schema in the process (in memory). As for producing, RIG again retrieves and caches the schemas used for serializing events.
```

Also, I've wondered, do we support Avro without also using the schema registry? Avro doesn't really require it..

Author: Not right now, currently we rely on Schema registry, we would have to add support for other type of storing schema such as files on file system.

Author: If we want to support Avro without schema registry, I would do it iteratively.

Reviewer: I wasn't thinking about storing the schema anywhere - instead, without a schema registry, the schema would be included in the message itself, for each message, so caching wouldn't even be appropriate.
But I agree, we should do this once the need arises.

### Producer details

- producer evaluates if serializing is turned on by checking `KAFKA_SERIALIZER` environment variable and if it's value is `avro`
- If it is, creates headers for Kafka event by appending `cloudEvents_` prefix for every field besides `data` field

>   - **for deep nested values we are using query encoding** since Kafka headers don't support nested values

Reviewer: perhaps you should point out that this only applies to so-called CloudEvents "extensions"

Author: Do we want to try stringified values instead of this monster?
